go-controller/pkg/node/gateway_localnet.go

```diff
@@ -107,55 +123,50 @@ func (n *OvnNode) initLocalnetGateway(subnet *net.IPNet, nodeAnnotator kube.Anno
 		ChassisID:      chassisID,
 		InterfaceID:    ifaceID,
 		MACAddress:     macAddress,
-		IPAddresses:    []*net.IPNet{gatewayIPCIDR},
-		NextHops:       []net.IP{gatewayNextHop},
+		IPAddresses:    gatewayIfAddrs,
+		NextHops:       nextHops,
 		NodePortEnable: config.Gateway.NodeportEnable,
 	})
 	if err != nil {
 		return err
 	}
 
-	if utilnet.IsIPv6CIDR(subnet) {
-		// TODO - IPv6 hack ... for some reason neighbor discovery isn't working here, so hard code a
-		// MAC binding for the gateway IP address for now - need to debug this further
-		err = util.LinkNeighAdd(link, gatewayIP, macAddress)
-		if err == nil {
-			klog.Infof("Added MAC binding for %s on %s", gatewayIP, localnetGatewayNextHopPort)
-		} else {
-			klog.Errorf("Error in adding MAC binding for %s on %s: %v", gatewayIP, localnetGatewayNextHopPort, err)
+	for _, ifaddr := range gatewayIfAddrs {
+		err = initLocalGatewayNATRules(localnetGatewayNextHopPort, ifaddr.IP)
+		if err != nil {
+			return fmt.Errorf("failed to add NAT rules for localnet gateway (%v)", err)
 		}
 	}
 
-	err = initLocalGatewayNATRules(localnetGatewayNextHopPort, gatewayIP)
-	if err != nil {
-		return fmt.Errorf("failed to add NAT rules for localnet gateway (%v)", err)
-	}
-
 	if config.Gateway.NodeportEnable {
 		localAddrSet, err := getLocalAddrs()
 		if err != nil {
 			return err
 		}
 		err = n.watchLocalPorts(
-			newLocalPortWatcherData(gatewayIP, n.recorder, localAddrSet),
+			newLocalPortWatcherData(gatewayIfAddrs, n.recorder, localAddrSet),
 		)
 		if err != nil {
 			return err
 		}
 	}
 
-	return err
+	return nil
 }
 
 type localPortWatcherData struct {
 	recorder     record.EventRecorder
-	gatewayIP    string
+	gatewayIPs   map[bool]string // maps from IsIPv6 to gateway IP string
```

Comment:
just curious, why is it not sufficient to add a `v6gatewayIP` and rename existing `gatewayIP` to `v4gatewayIP`? Isn't that cleaner?

Reply:
doing it this way lets us say:

    gatewayIP := npw.gatewayIPs[isIPv6Service]

rather than

    var gatewayIP string
    if isIPv6Service {
          gatewayIP = npw.gatewayIPv6
    } else {
          gatewayIP = npw.gatewayIPv4
    }

etc. Though I agree that the `map[bool]` is an ugly way to represent that, and I guess we're not doing that anywhere else...